lib/action_controller/serialization.rb

```diff
@@ -25,6 +25,11 @@ def get_serializer(resource, options = {})
           "Please pass 'adapter: false' or see ActiveSupport::SerializableResource#serialize"
         options[:adapter] = false
       end
+      if resource.respond_to?(:current_page) && resource.respond_to?(:total_pages)
+        options[:pagination] = {}
+        options[:pagination][:original_url] = original_url
+        options[:pagination][:query_parameters] = query_parameters
+      end
```

Comment:
could we maybe instead pass in a `context` so that serialization isn't as tightly coupled to the controller?

e.g. in the controller, just

```ruby
options.fetch(:context) { options[:context] = request }
```
Then in the ArraySerializer

```ruby
class ArraySerializer
   def object # Same interface as Serializer
     @resource
   end
end
```

Then in the JsonApi adapter

```ruby
  def serializable_hash(options = nil)
          options ||= {}
          if serializer.respond_to?(:each)
            serializer.each do |s|
               # code
            end
            add_links(options)
          end

        private

        def add_links(options)
          links = @hash[:links] ||= {} # no reason for them to be nil or false
          @hash[:links].update(JsonApi::PaginationLinks.new(object, links).serializable_hash(options))
        end
end
```

Then in the [PaginationLinks builder](https://github.com/rails-api/active_model_serializers/pull/1041#discussion_r36998222)

```ruby
def initialize(collection, links)
  @collection = collection
end
def serializable_hash(options)
  pagination_links = {}
  return pagination_links if !collection.respond_to?(:current_page) && !collection.respond_to?(:total_pages)
  return pagination_links if collection.total_pages == FIRST_PAGE

  context = options.fetch(:context)
  original_url = context.original_url
  url_base = original_url[/\A[^?]+/]
  query_parameters = context.query_parameters.dup || original_url[/\?.*/] || {}

  current_links = links
  self_link = current_links.fetch(:self) { original_url }

  collection_size            = collection.size
  collection_current_page    = collection.current_page
  collection_total_pages     = collection.total_pages


  params = query_parameters.merge!(page: {})
  link_template = "#{url_base}?%s"
  # only add first/previous page links when not on the first page (i.e. we can go back)
  if collection_current_page != FIRST_PAGE
    pagination_links[:first] = link_template % params[:page].merge(size: collection_size, number: FIRST_PAGE).to_query
    pagination_links[:prev]  = link_template % params[:page].merge(size: collection_size, number: (collection_current_page - FIRST_PAGE)).to_query
  end

  # only add next/last page links when not on the last page (i.e. we can go forward)
  if collection_current_page != collection_total_pages
    pagination_links[:next] = link_template % params[:page].merge(size: collection_size, number: (collection_current_page + FIRST_PAGE)).to_query
    pagination_links[:last] = link_template % params[:page].merge(size: collection_size, number: collection_total_pages).to_query
  end
  pagination_links
end
```